### PR TITLE
Unique IPs

### PR DESCRIPTION
**Pull request by james-antill**
Still super beta, but kind of works and everyone can look at what is here and what the problems are etc.

Hasn't run though black/etc.

### CODE REVIEW

**Comment by nphilipp** [comment]
Resolved a conflict in the latest push

**Comment by james-antill** [comment]
Hey, @LenkaSeg I'm not sure how to fix the mock connection problem that's failing the tests ... you have time on Monday to look at it together?

File: mirrors_countme/totals.py [hunk 1]
```diff
@@ -118,38 +118,110 @@ class RawDB(SQLiteReader):
...
-            f"SELECT {item_select}"
+
+        # So this is a "problem" ... we could do a GROUP BY as we do in
+        # SplitWeekDays(), but that is basically doing all the same work.
+        # Just counting the rows gets us to roughly 8:1 ... so go with that.
+        cursor = self._connection.execute(
+            f"SELECT COUNT(*)"
             f" FROM {self._tablename}"
-            f" WHERE timestamp >= {start_ts} AND timestamp < {end_ts}"
+            f" WHERE timestamp >= {start_ts} AND timestamp < {end_ts} AND sys_age < 0"
```
Comment (nphilipp) [inline mirrors_countme/totals.py:185]: Not sure about this – shouldn’t unique IP statistics be about all log entries (including those with `countme`?

File: mirrors_countme/totals.py [hunk 2]
```diff
@@ -118,38 +118,110 @@ class RawDB(SQLiteReader):
...
+        item_select = ",".join(self.select)
+        start_ts = self.start_ts
+
+        for d in range(7):
+            end_ts = start_ts + (WEEK_LEN / 7)
+            # Can add "COUNT(*) as nip" to test, but need to remove it for writes
+            cursor = self.rawdb._connection.execute(
+                f"SELECT {item_select}"
+                f" FROM {self.rawdb._tablename}"
+                f" WHERE timestamp >= {start_ts} AND timestamp < {end_ts} AND sys_age < 0"
```
Comment (nphilipp) [inline mirrors_countme/totals.py:219]: Same thing (with or without `countme`).

File: mirrors_countme/totals.py [hunk 3]
```diff
@@ -118,38 +118,110 @@ class RawDB(SQLiteReader):
...
+    def __init__(self, fp, **kwargs):
+        super().__init__(fp, **kwargs)
+
+    @property
+    def mintime(self):
+        return self.mintime_unique
+
+    @property
+    def maxtime(self):
+        return self.maxtime_unique
```
Comment (nphilipp) [inline mirrors_countme/totals.py:169]: Depending on if we want to count `countme`, this probably should use `mintime`, `maxtime` properties which cover all log entries (i.e. probably requires that we have a base class which doesn’t override `mintime`, `maxtime` like `RawDB` does above).

**Review by nphilipp — COMMENTED** [review]
We need to decide if the unique IP statistics should include `countme` log entries – I think they should.

**Comment by james-antill** [comment]
Yeh, I didn't even think about counting the "countme" entries as well ... mentally split them into seperate things. I guess we should.

**Comment by james-antill** [comment]
The mintime one is more problematic ... because when trimming or processing the total() statistics we need to know "what is the oldest non-countme entry".

**Comment by nphilipp** [comment]
> The mintime one is more problematic ... because when trimming or processing the total() statistics we need to know "what is the oldest non-countme entry".

Not sure what you mean… when trimming unique IPs only, the `unique` versions of `mintime`, `maxtime` are used, when processing into totals, `RawDBU` should count in entries both with and without `countme`. What am I missing? 😅

**Comment by james-antill** [comment]
With both totals()/trim the question you are asking the DB is "what is the timestamp of the oldest week that has unique IP data" ... and even if we count the countme data as part of unique IP data we could easily have say three months of countme data but only one month of unique IP data, and you need the answer to come back as one month in both cases.

**Comment by nphilipp** [comment]
> With both totals()/trim the question you are asking the DB is "what is the timestamp of the oldest week that has unique IP data" ... and even if we count the countme data as part of unique IP data we could easily have say three months of countme data but only one month of unique IP data, and you need the answer to come back as one month in both cases.

Ahh, I think I get it now, we need to start looking at the last week with unique IP data in `totals.db`, I somehow confused this with rows in `raw.db`…

Anyway, I noticed another issue – so far the unique IP code inserted entries for all different values of `sys_age` into `totals.db` (again in the case of `countme` present), my latest push should fix this. Would you please review? If the code as is works, I would unite the commits that belong together and merge.

**Comment by james-antill** [comment]
Your last four commits seem good to me.
I would probably have tried altering the sys_age data to be -1 before it passed to write_item(), but the BucketSelectUniqueIP() thing seems cleaner :).

**Comment by nphilipp** [comment]
> I would probably have tried altering the sys_age data to be -1 before it passed to write_item()

I’m not sure if this would have worked, `BucketSelect` is a tuple, i.e. immutable.

**Review by nphilipp — APPROVED** [review]

**Comment by nphilipp** [comment]
Latest push cleans up the commits, I’ll merge.

**Pull request merged**